src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -84,25 +100,75 @@ private StreamingClientProvider() {
    * @return A streaming client
    */
   public SnowflakeStreamingIngestClient getClient(Map<String, String> connectorConfig) {
+    SnowflakeStreamingIngestClient resultClient = null;
+
     if (Boolean.parseBoolean(
         connectorConfig.getOrDefault(
             SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG,
             Boolean.toString(ENABLE_STREAMING_CLIENT_OPTIMIZATION_DEFAULT)))) {
-      LOGGER.info(
-          "Streaming client optimization is enabled, returning the existing streaming client if"
-              + " valid");
-      this.providerLock.lock();
-      // recreate streaming client if needed
-      if (!StreamingClientHandler.isClientValid(this.parameterEnabledClient)) {
-        LOGGER.error("Current streaming client is invalid, recreating client");
-        this.parameterEnabledClient = this.streamingClientHandler.createClient(connectorConfig);
+      LOGGER.debug(
+          "Streaming client optimization is enabled per worker node. Reusing valid clients when"
+              + " possible");
+
+      try {
+        this.providerLock.lock();
+
+        // remove invalid clients
+        List<Map.Entry<Properties, SnowflakeStreamingIngestClient>> invalidClientProps =
+            this.registeredClientMap.entrySet().stream()
+                .filter(entry -> !StreamingClientHandler.isClientValid(entry.getValue()))
+                .collect(Collectors.toList());
+        if (!invalidClientProps.isEmpty()) {
+          List<String> invalidClientNames = new ArrayList<>();
+
+          for (Map.Entry<Properties, SnowflakeStreamingIngestClient> entry : invalidClientProps) {
+            SnowflakeStreamingIngestClient invalidClient = entry.getValue();
+            invalidClientNames.add(
+                invalidClient != null
+                        && invalidClient.getName() != null
+                        && !invalidClient.getName().isEmpty()
+                    ? invalidClient.getName()
+                    : "noClientNameFound");
+            this.registeredClientMap.remove(entry.getKey());
+          }
+
+          LOGGER.error(
+              "Found and removed {} invalid clients: {}",
+              invalidClientProps.size(),
+              invalidClientNames.toString());
+        }
+
+        // look for client corresponding to the input properties or create new client
+        Properties inputProps = StreamingClientHandler.getClientProperties(connectorConfig);
+        if (this.registeredClientMap.containsKey(inputProps)) {
+          resultClient = this.registeredClientMap.get(inputProps);
+          LOGGER.debug("Using existing streaming client with name: {}", resultClient.getName());
```

Comment:
Should we update this and the one below to INFO? Otherwise it would be hard if we have issues in this area in the future

Reply:
yeah thats true, updating this and the else statement log level to info